# Benchmark de Estrategias en Múltiples Datasets

## Objetivo del Notebook

En este notebook quiero validar si las estrategias típicas de balanceo funcionan igual en todos los escenarios o si dependen del dataset. Voy a probar 3 estrategias (Base, Pesos 'balanced' y SMOTE) sobre 4 datasets con distintos niveles de desbalanceo (desde el alemán que es pequeño hasta el de fraude de tarjetas que es masivo).

La idea es tener una visión general antes de meterme a optimizar modelos complejos.

**Estrategias a Comparar:**
1.  **Estrategia 1: Convencional** (Línea Base).
2.  **Estrategia 2: Coste Sensible Fijo** (Usando `class_weight='balanced'`).
3.  **Estrategia 3: Balanceo de Datos** (Usando `SMOTE`).

**Métricas Clave:**
* **`Balanced Accuracy`:** Mide el equilibrio general del modelo.
* **`F1-Score (Clase Minoritaria)`:** Mide la efectividad (equilibrio Precisión/Recall) para detectar la clase "Mal Riesgo" o "Fraude", que es nuestro objetivo principal.

In [ ]:
import pandas as pd
import sys
import os
from sklearn.datasets import fetch_openml

# Modelos a comparar
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier

# Métricas y utilidades
from sklearn.metrics import confusion_matrix, balanced_accuracy_score, f1_score
from IPython.display import display

# Mis funciones del módulo src
project_root = os.path.abspath(os.path.join(os.getcwd(), '..'))
if project_root not in sys.path:
    sys.path.append(project_root)

from src.benchmark_utils import evaluate_models, evaluate_cost_sensitive_models, evaluate_smote_models

## 1. Definición de Modelos a Probar

Se seleccionan tres tipos de modelos representativos: Regresión Logística (lineal), Árbol de Decisión (árbol simple) y Random Forest (ensamble de árboles).

**Nota sobre SVC (Support Vector Classifier):**
Aunque `SVC` fue importado en la celda anterior, se ha **comentado y excluido** de este benchmark (`models_to_test`).

La razón es su **alto coste computacional**. El tiempo de entrenamiento de SVC escala de forma cuadrática o cúbica ($O(n^2)$ a $O(n^3)$) con el número de muestras. Ejecutarlo en datasets grandes como "Credit Card Fraud" (284k muestras) o "Give Me Some Credit" (150k muestras) bajo 3 estrategias diferentes (36* ejecuciones) haría que el notebook tardara horas en completarse.

El `SVC` sí se analiza en profundidad en el notebook `modelos_convencionales_german_credit_data.ipynb`, que utiliza un dataset más pequeño (1000 muestras) donde su coste es manejable.

In [ ]:
models_to_test = {
    'Logistic Regression': LogisticRegression(random_state=42, max_iter=2000),
    'Random Forest': RandomForestClassifier(random_state=42),
    #'SVC': SVC(random_state=42),
    'Decision Tree': DecisionTreeClassifier(random_state=42)
}

# Diccionario de datasets (con "Give Me Some Credit" activado)
datasets_to_test = {
    'German Credit': {'source': 'openml', 'id': 31, 'target_col': 'class'},
    'Taiwan Credit Default': {'source': 'openml', 'id': 42477, 'target_col': 'class'},
    'Credit Card Fraud': {'source': 'openml', 'id': 1597, 'target_col': 'class'},
    'Give Me Some Credit': {'source': 'csv', 'path': '../data/cs-training.csv', 'target_col': 'SeriousDlqin2yrs'}
}

### 1. Estrategia Base (Sin balanceo)

Se definen los 4 datasets a probar: German Credit, Taiwan Credit Default, Credit Card Fraud y Give Me Some Credit.

Un paso clave en la carga de datos es **mapear la variable objetivo (y)** para que la **clase minoritaria (impago/fraude) sea siempre 1** y la mayoritaria sea 0. Esto estandariza la evaluación.

A continuación, se ejecuta el benchmark convencional.

In [ ]:
# Lista para guardar los resultados convencionales
all_results_conv = []

print("INICIANDO BENCHMARK (1. CONVENCIONAL)")

for name, info in datasets_to_test.items():
    try:
        print(f"\n Dataset: {name}")
        
        # Carga de datos
        if info['source'] == 'openml':
            data = fetch_openml(data_id=info['id'], as_frame=True, parser='auto')
            X = data.data
            y_raw = data.target if 'target' in data else data['class']

            # Ajustamos las etiquetas para que minoritaria = 1
            y_str = y_raw.astype(str)
            counts = y_str.value_counts()

            if len(counts) > 2:
                print(f"Aviso: El dataset {name} tiene más de 2 clases.")
                y = pd.Series(pd.factorize(y_raw)[0], index=y_raw.index)
            else:
                minority_label = counts.idxmin()
                majority_label = counts.idxmax()
                
                print(f"Mapeando: '{majority_label}' (Mayoritaria) -> 0, '{minority_label}' (Minoritaria) -> 1")
                y = y_str.map({majority_label: 0, minority_label: 1}).astype(int)
                y.index = y_raw.index # Preservar el índice


        elif info['source'] == 'csv':
            df = pd.read_csv(info['path'])
            # Asumimos que la primera columna (Unnamed: 0) es un índice y la quitamos
            if 'Unnamed: 0' in df.columns:
                df = df.drop(columns=['Unnamed: 0'])
            X = df.drop(columns=[info['target_col']])
            y_raw = df[info['target_col']]
            # El dataset CSV tiene valores nulos, los rellenamos con la media
            X = X.fillna(X.mean())

            # Este dataset ya es 0 (bueno) y 1 (malo). 
            y = y_raw.astype(int)

        
        # Guardar los datos para las siguientes ejecuciones
        info['X_data'] = X
        info['y_data'] = y

        results = evaluate_models(X, y, models_to_test, name)
        all_results_conv.extend(results)
        print(f"Evaluación convencional para {name} completada.")

    except Exception as e:
        print(f"No se pudo procesar el dataset {name}. Error: {e}")

# Resultados
final_results_conv_df = pd.DataFrame(all_results_conv)
print("\n--- TABLA COMPARATIVA (1. Convencional) ---")
display(final_results_conv_df.sort_values(by=['dataset', 'balanced_accuracy'], ascending=[True, False]))

INICIANDO BENCHMARK (1. CONVENCIONAL)

--- Cargando y procesando dataset: German Credit ---
Mapeando: 'good' (Mayoritaria) -> 0, 'bad' (Minoritaria) -> 1
--- Evaluando Dataset: German Credit ---

Probando modelo: Logistic Regression...
Matriz de Confusión:
[[187  23]
 [ 42  48]]
Balanced Accuracy: 0.712
F1-Score (clase 'mala'): 0.596

Probando modelo: Random Forest...
Matriz de Confusión:
[[196  14]
 [ 55  35]]
Balanced Accuracy: 0.661
F1-Score (clase 'mala'): 0.504

Probando modelo: Decision Tree...
Matriz de Confusión:
[[158  52]
 [ 51  39]]
Balanced Accuracy: 0.593
F1-Score (clase 'mala'): 0.431
Evaluación convencional para German Credit completada.

--- Cargando y procesando dataset: Taiwan Credit Default ---
Mapeando: '0' (Mayoritaria) -> 0, '1' (Minoritaria) -> 1
--- Evaluando Dataset: Taiwan Credit Default ---

Probando modelo: Logistic Regression...
Matriz de Confusión:
[[6805  204]
 [1521  470]]
Balanced Accuracy: 0.603
F1-Score (clase 'mala'): 0.353

Probando modelo: Random F

dataset                model  balanced_accuracy  \
7       Credit Card Fraud        Random Forest           0.881727   
8       Credit Card Fraud        Decision Tree           0.868091   
6       Credit Card Fraud  Logistic Regression           0.807345   
0           German Credit  Logistic Regression           0.711905   
1           German Credit        Random Forest           0.661111   
2           German Credit        Decision Tree           0.592857   
11    Give Me Some Credit        Decision Tree           0.615006   
10    Give Me Some Credit        Random Forest           0.587051   
9     Give Me Some Credit  Logistic Regression           0.519659   
4   Taiwan Credit Default        Random Forest           0.651745   
5   Taiwan Credit Default        Decision Tree           0.608991   
3   Taiwan Credit Default  Logistic Regression           0.603478   

    f1_score_bad_class             Tipo  
7             0.849624  1. Convencional  
8             0.770318  1. Convencional  
6             0.716535  1. Convencional  
0             0.596273  1. Convencional  
1             0.503597  1. Convencional  
2             0.430939  1. Convencional  
11            0.276404  1. Convencional  
10            0.277500  1. Convencional  
9             0.077471  1. Convencional  
4             0.461736  1. Convencional  
5             0.392377  1. Convencional  
3             0.352720  1. Convencional

### 2.1. Resultados del Benchmark Convencional

La tabla de resultados y la de promedios demuestran el problema central:

* **Fracaso Total en Desbalanceo Extremo:** En "Give Me Some Credit" (6.7% de impagos), la Regresión Logística convencional obtiene un **F1-Score de 0.077**. Esto es un fracaso total: el modelo no tiene ninguna capacidad real de detectar impagos.
* **Precisión Engañosa:** En "Credit Card Fraud" (0.17% de fraude), los modelos convencionales (especialmente RF y DT) obtienen un `Balanced Accuracy` y un `F1-Score` altos (ej. RF F1-Score 0.850). Esto se debe a que el conjunto de prueba es muy pequeño y los modelos "tuvieron suerte", pero esta métrica es engañosa, como veremos.
* **Línea Base Promedio:** En promedio, el Random Forest (Bal. Acc. 0.695) es el mejor modelo convencional.

In [ ]:
print(" RENDIMIENTO MEDIO POR MODELO CONVENCIONAL (TODOS LOS DATASETS) ")

# Agrupamos por 'model' y calculamos la media de las métricas
average_performance = final_results_conv_df.groupby('model')[['balanced_accuracy', 'f1_score_bad_class']].mean()

# Ordenamos por la métrica que consideremos más importante para ver el ranking
average_performance_sorted = average_performance.sort_values(by='balanced_accuracy', ascending=False)

display(average_performance_sorted)

--- RENDIMIENTO MEDIO POR MODELO CONVENCIONAL (TODOS LOS DATASETS) ---


balanced_accuracy  f1_score_bad_class
model                                                     
Random Forest                 0.695409            0.523114
Decision Tree                 0.671236            0.467510
Logistic Regression           0.660597            0.435750

### 2. Estrategias de Balanceo (Pesos y SMOTE)

Ahora pruebo dos enfoques para intentar arreglar el desastre de los modelos lineales en datasets desbalanceados:
1.  **Cost Sensitive:** Pasar `class_weight='balanced'` al modelo.
2.  **SMOTE:** Crear datos sintéticos en el train.

Uso las funciones `evaluate_cost_sensitive_models` y `evaluate_smote_models` que se recogen dentro de `benchmark_utils.py` en la carpeta src.

In [ ]:
all_results_cs = []
print("\nINICIANDO BENCHMARK (2. COSTE SENSIBLE)")

# Este bucle ahora REUTILIZA los datos X e y cargados en la celda anterior
for name, info in datasets_to_test.items():
    try:
        if 'X_data' in info:
            print(f"\nProcesando {name} (Coste Sensible)...")
            # Llamar a la función de evaluación de COSTE SENSIBLE (celda [9])
            results_cs = evaluate_cost_sensitive_models(info['X_data'], info['y_data'], models_to_test, name)
            all_results_cs.extend(results_cs)
            print(f"Evaluación CSL para {name} completada.")
        
        else:
            print(f"Saltando {name} (no se cargaron los datos en el paso anterior).")

    except Exception as e:
        print(f"No se pudo procesar el dataset {name} (CSL). Error: {e}")

final_results_cs_df = pd.DataFrame(all_results_cs)
print("\n--- TABLA COMPARATIVA (2. Coste Sensible) ---")
display(final_results_cs_df.sort_values(by=['dataset', 'balanced_accuracy'], ascending=[True, False]))


INICIANDO BENCHMARK (2. COSTE SENSIBLE)

Procesando German Credit (Coste Sensible)...
--- Evaluando Dataset (Sensible a Costes): German Credit ---

Probando modelo: Logistic Regression (Sensible al Coste)...
Matriz de Confusión:
[[150  60]
 [ 21  69]]
Balanced Accuracy: 0.740
F1-Score (clase 'mala'): 0.630
              precision    recall  f1-score   support

     Clase 0       0.88      0.71      0.79       210
     Clase 1       0.53      0.77      0.63        90

    accuracy                           0.73       300
   macro avg       0.71      0.74      0.71       300
weighted avg       0.77      0.73      0.74       300


Probando modelo: Random Forest (Sensible al Coste)...
Matriz de Confusión:
[[194  16]
 [ 61  29]]
Balanced Accuracy: 0.623
F1-Score (clase 'mala'): 0.430
              precision    recall  f1-score   support

     Clase 0       0.76      0.92      0.83       210
     Clase 1       0.64      0.32      0.43        90

    accuracy                           0.74  

dataset                model  balanced_accuracy  \
6       Credit Card Fraud  Logistic Regression           0.924743   
7       Credit Card Fraud        Random Forest           0.851334   
8       Credit Card Fraud        Decision Tree           0.824137   
0           German Credit  Logistic Regression           0.740476   
1           German Credit        Random Forest           0.623016   
2           German Credit        Decision Tree           0.591270   
9     Give Me Some Credit  Logistic Regression           0.721863   
11    Give Me Some Credit        Decision Tree           0.598272   
10    Give Me Some Credit        Random Forest           0.572509   
3   Taiwan Credit Default  Logistic Regression           0.663928   
4   Taiwan Credit Default        Random Forest           0.645418   
5   Taiwan Credit Default        Decision Tree           0.618091   

    f1_score_bad_class                              Tipo  
6             0.119169  2. Coste Sensible (class_weight)  
7             0.815686  2. Coste Sensible (class_weight)  
8             0.695652  2. Coste Sensible (class_weight)  
0             0.630137  2. Coste Sensible (class_weight)  
1             0.429630  2. Coste Sensible (class_weight)  
2             0.432432  2. Coste Sensible (class_weight)  
9             0.282618  2. Coste Sensible (class_weight)  
11            0.252240  2. Coste Sensible (class_weight)  
10            0.241678  2. Coste Sensible (class_weight)  
3             0.467838  2. Coste Sensible (class_weight)  
4             0.449313  2. Coste Sensible (class_weight)  
5             0.405127  2. Coste Sensible (class_weight)

### 3. Resultados del Benchmark CSL (`class_weight='balanced'`)

Los resultados de CSL revelan un patrón fundamental:

* **Mejora en Desbalanceo Moderado:** En "Give Me Some Credit" (6.7%), el CSL mejora drásticamente el F1-Score de la Regresión Logística (de 0.077 a **0.283**).
* **Fracaso en Desbalanceo Extremo:** En "Credit Card Fraud" (0.17%), el CSL (`'balanced'`) es **desastroso**. El F1-Score de la Regresión Logística colapsa de 0.717 (convencional) a **0.119**.
* **Fallo en Random Forest:** La `Balanced Accuracy` promedio de RF *empeora* (de 0.695 a 0.673).

**Conclusión clave:** `class_weight='balanced'` funciona para modelos lineales en desbalanceo *moderado*, pero **falla estrepitosamente** en desbalanceo *extremo* (colapsa el F1-Score) y no funciona bien con Random Forest.

In [ ]:
print(" RENDIMIENTO MEDIO POR MODELO COSTE (TODOS LOS DATASETS)")

# Agrupamos por 'model' y calculamos la media de las métricas
average_performance = final_results_cs_df.groupby('model')[['balanced_accuracy', 'f1_score_bad_class']].mean()

# Ordenamos por la métrica que consideremos más importante para ver el ranking
average_performance_sorted = average_performance.sort_values(by='balanced_accuracy', ascending=False)

display(average_performance_sorted)

📊 --- RENDIMIENTO MEDIO POR MODELO COSTE (TODOS LOS DATASETS) --- 📊


balanced_accuracy  f1_score_bad_class
model                                                     
Logistic Regression           0.762753            0.374941
Random Forest                 0.673069            0.484077
Decision Tree                 0.657942            0.446363

In [ ]:
all_results_smote = []
print("\nINICIANDO BENCHMARK (3. BALANCEO SMOTE)")

for name, info in datasets_to_test.items():
    try:
        if 'X_data' in info: 
            print(f"\nProcesando {name} (SMOTE)...")            
            results_smote = evaluate_smote_models(info['X_data'], info['y_data'], models_to_test, name)
            all_results_smote.extend(results_smote)
            print(f"Evaluación SMOTE para {name} completada.")
        
        else:
            print(f"Saltando {name} (no se cargaron los datos).")
            
    except Exception as e:
        print(f"No se pudo procesar el dataset {name} (SMOTE). Error: {e}")

final_results_smote_df = pd.DataFrame(all_results_smote)
print("\n--- TABLA COMPARATIVA (3. Balanceo SMOTE) --- ")
display(final_results_smote_df.sort_values(by=['dataset', 'balanced_accuracy'], ascending=[True, False]))


INICIANDO BENCHMARK (3. BALANCEO SMOTE)

Procesando German Credit (SMOTE)...
--- Evaluando Dataset (Balanceo SMOTE): German Credit ---
Evaluación SMOTE para German Credit completada.

Procesando Taiwan Credit Default (SMOTE)...
--- Evaluando Dataset (Balanceo SMOTE): Taiwan Credit Default ---
Evaluación SMOTE para Taiwan Credit Default completada.

Procesando Credit Card Fraud (SMOTE)...
--- Evaluando Dataset (Balanceo SMOTE): Credit Card Fraud ---
Evaluación SMOTE para Credit Card Fraud completada.

Procesando Give Me Some Credit (SMOTE)...
--- Evaluando Dataset (Balanceo SMOTE): Give Me Some Credit ---
Evaluación SMOTE para Give Me Some Credit completada.

--- TABLA COMPARATIVA (3. Balanceo SMOTE) --- 


dataset                model  balanced_accuracy  \
6       Credit Card Fraud  Logistic Regression           0.930873   
7       Credit Card Fraud        Random Forest           0.891804   
8       Credit Card Fraud        Decision Tree           0.873927   
0           German Credit  Logistic Regression           0.740476   
1           German Credit        Random Forest           0.667460   
2           German Credit        Decision Tree           0.580952   
9     Give Me Some Credit  Logistic Regression           0.716450   
10    Give Me Some Credit        Random Forest           0.647169   
11    Give Me Some Credit        Decision Tree           0.622387   
4   Taiwan Credit Default        Random Forest           0.676932   
3   Taiwan Credit Default  Logistic Regression           0.660037   
5   Taiwan Credit Default        Decision Tree           0.616844   

    f1_score_bad_class                 Tipo  
6             0.115215  3. Balanceo (SMOTE)  
7             0.831541  3. Balanceo (SMOTE)  
8             0.502262  3. Balanceo (SMOTE)  
0             0.631579  3. Balanceo (SMOTE)  
1             0.522876  3. Balanceo (SMOTE)  
2             0.409091  3. Balanceo (SMOTE)  
9             0.268322  3. Balanceo (SMOTE)  
10            0.352529  3. Balanceo (SMOTE)  
11            0.250965  3. Balanceo (SMOTE)  
4             0.501361  3. Balanceo (SMOTE)  
3             0.462741  3. Balanceo (SMOTE)  
5             0.408163  3. Balanceo (SMOTE)

In [ ]:
print("RENDIMIENTO MEDIO POR MODELO SMOTE (TODOS LOS DATASETS) ")

# Agrupamos por 'model' y calculamos la media de las métricas
average_performance = final_results_smote_df.groupby('model')[['balanced_accuracy', 'f1_score_bad_class']].mean()

# Ordenamos por la métrica que consideremos más importante para ver el ranking
average_performance_sorted = average_performance.sort_values(by='balanced_accuracy', ascending=False)

display(average_performance_sorted)

--- RENDIMIENTO MEDIO POR MODELO COSTE (TODOS LOS DATASETS) --- 


balanced_accuracy  f1_score_bad_class
model                                                     
Logistic Regression           0.761959            0.369464
Random Forest                 0.720841            0.552077
Decision Tree                 0.673528            0.392620

### 4. Resultados del Benchmark SMOTE

Los resultados de SMOTE (visibles en la tabla de la celda 13) muestran un patrón diferente:

* **Éxito en Regresión Logística:** Al igual que CSL, SMOTE **mejora drásticamente** el rendimiento de LR (promedio de 0.661 a 0.762).
* **Éxito en Random Forest:** A diferencia de CSL, SMOTE **sí mejora** el rendimiento de RF (promedio de 0.695 a 0.720).

Esto sugiere que los modelos de Random Forest se benefician más de ver un conjunto de datos balanceado (SMOTE) que de una penalización interna (CSL).

In [ ]:
final_results_conv_df['Tipo'] = '1. Convencional'
final_results_cs_df['Tipo'] = '2. Coste Sensible (class_weight)'
final_results_smote_df['Tipo'] = '3. Balanceo (SMOTE)' 

# Unimos las TRES tablas de resultados
comparison_df = pd.concat([final_results_conv_df, final_results_cs_df, final_results_smote_df]) 

# Mostramos la tabla final, ordenada para facilitar la comparación
print("--- TABLA COMPARATIVA FINAL: ESTRATEGIAS vs. DATASETS ---")
display(comparison_df.sort_values(by=['dataset', 'model', 'Tipo']))

--- TABLA COMPARATIVA FINAL: ESTRATEGIAS vs. DATASETS ---


dataset                model  balanced_accuracy  \
8       Credit Card Fraud        Decision Tree           0.868091   
8       Credit Card Fraud        Decision Tree           0.824137   
8       Credit Card Fraud        Decision Tree           0.873927   
6       Credit Card Fraud  Logistic Regression           0.807345   
6       Credit Card Fraud  Logistic Regression           0.924743   
6       Credit Card Fraud  Logistic Regression           0.930873   
7       Credit Card Fraud        Random Forest           0.881727   
7       Credit Card Fraud        Random Forest           0.851334   
7       Credit Card Fraud        Random Forest           0.891804   
2           German Credit        Decision Tree           0.592857   
2           German Credit        Decision Tree           0.591270   
2           German Credit        Decision Tree           0.580952   
0           German Credit  Logistic Regression           0.711905   
0           German Credit  Logistic Regression           0.740476   
0           German Credit  Logistic Regression           0.740476   
1           German Credit        Random Forest           0.661111   
1           German Credit        Random Forest           0.623016   
1           German Credit        Random Forest           0.667460   
11    Give Me Some Credit        Decision Tree           0.615006   
11    Give Me Some Credit        Decision Tree           0.598272   
11    Give Me Some Credit        Decision Tree           0.622387   
9     Give Me Some Credit  Logistic Regression           0.519659   
9     Give Me Some Credit  Logistic Regression           0.721863   
9     Give Me Some Credit  Logistic Regression           0.716450   
10    Give Me Some Credit        Random Forest           0.587051   
10    Give Me Some Credit        Random Forest           0.572509   
10    Give Me Some Credit        Random Forest           0.647169   
5   Taiwan Credit Default        Decision Tree           0.608991   
5   Taiwan Credit Default        Decision Tree           0.618091   
5   Taiwan Credit Default        Decision Tree           0.616844   
3   Taiwan Credit Default  Logistic Regression           0.603478   
3   Taiwan Credit Default  Logistic Regression           0.663928   
3   Taiwan Credit Default  Logistic Regression           0.660037   
4   Taiwan Credit Default        Random Forest           0.651745   
4   Taiwan Credit Default        Random Forest           0.645418   
4   Taiwan Credit Default        Random Forest           0.676932   

    f1_score_bad_class                              Tipo  
8             0.770318                   1. Convencional  
8             0.695652  2. Coste Sensible (class_weight)  
8             0.502262               3. Balanceo (SMOTE)  
6             0.716535                   1. Convencional  
6             0.119169  2. Coste Sensible (class_weight)  
6             0.115215               3. Balanceo (SMOTE)  
7             0.849624                   1. Convencional  
7             0.815686  2. Coste Sensible (class_weight)  
7             0.831541               3. Balanceo (SMOTE)  
2             0.430939                   1. Convencional  
2             0.432432  2. Coste Sensible (class_weight)  
2             0.409091               3. Balanceo (SMOTE)  
0             0.596273                   1. Convencional  
0             0.630137  2. Coste Sensible (class_weight)  
0             0.631579               3. Balanceo (SMOTE)  
1             0.503597                   1. Convencional  
1             0.429630  2. Coste Sensible (class_weight)  
1             0.522876               3. Balanceo (SMOTE)  
11            0.276404                   1. Convencional  
11            0.252240  2. Coste Sensible (class_weight)  
11            0.250965               3. Balanceo (SMOTE)  
9             0.077471                   1. Convencional  
9             0.282618  2. Coste Sensible (class_weight)  
9             0.268322               3. Balanceo

## 5. Conclusiones Finales del Benchmark

La tabla comparativa final consolida los hallazgos:

1.  **Para Regresión Logística (Lineal):**
    * En desbalanceo **moderado** ("German Credit", "Taiwan"), tanto **CSL (`'balanced'`)** como **SMOTE** son estrategias excelentes y muy similares, mejorando tanto `Balanced Accuracy` como `F1-Score` (ej. German F1-Score: 0.596 -> **0.630**).
    * En desbalanceo **extremo** ("Credit Card Fraud"), ambas estrategias **fallan**, colapsando el `F1-Score` (0.717 -> **0.119** con CSL y **0.115** con SMOTE).

2.  **Para Random Forest (Bagging):**
    * **SMOTE** es la estrategia ganadora en promedio (mejora F1 de 0.523 a 0.552).
    * **CSL (`'balanced'`)** es una mala estrategia; empeora tanto la `Balanced Accuracy` como el `F1-Score` en promedio.

**Conclusión General:** Este notebook justifica la necesidad de estrategias más avanzadas. `class_weight='balanced'` y `SMOTE` (con parámetros por defecto) no son soluciones universales y **fallan en los casos de desbalanceo extremo**, que es donde se centra el notebook `modelos_avanzados.ipynb`.

### Conclusiones rápidas

1.  **Regresión Logística:** Necesita ayuda sí o sí. Tanto los pesos como SMOTE la reviven en datasets difíciles, aunque en el de fraude extremo sigue sufriendo.
2.  **Random Forest:** Curiosamente, ponerle `class_weight='balanced'` a veces empeora las cosas. SMOTE parece sentarle mejor en general.
3.  **No hay bala de plata:** Ninguna estrategia gana siempre. En el siguiente notebook (`modelos_avanzados`) me centraré en afinar esto con XGBoost y pesos manuales.